### PR TITLE
Todo comments implementation

### PR DESCRIPTION
**Pull request by Therzok**
Switches the todo comments implementation from the custom C# Parsed Document one to the roslyn engine.

### CODE REVIEW

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Editor.Extension/TagCommentsTextEditorExtension.cs [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+					return;
+				
+				var offset = Editor.LocationToOffset (todoItem.MappedLine, todoItem.MappedColumn);
+				var newTask = new QuickTask (todoItem.Message, offset, DiagnosticSeverity.Info);
+				newTasks.Add (newTask);
+			}
+
+			if (token.IsCancellationRequested)
+				return;
+			Application.Invoke ((o, args2) => {
```
Comment (mkrueger) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Editor.Extension/TagCommentsTextEditorExtension.cs]: Shouldn't we use Runtime.RunInMainThread instead ?

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Tasks/CommentTasksProvider.Legacy.cs [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+
+					// Discard files with known IToDoCommentService implementations
+					return mt != "text/x-csharp";
+				}).ToArray ();
+
+				Task.Run (async () => {
+					try {
+						await tags.UpdateAsync (project, files, token);
+					} catch (TaskCanceledException) {
+					} catch (AggregateException ae) {
```
Comment (mkrueger) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Tasks/CommentTasksProvider.Legacy.cs]: I'm pretty sure AggregateException can't be thrown here - no Wait() or Task.WhenAll etc. 

But what can be thrown is an OperationCanceledException which is not handled here. Never use TaskCanceledException.

CancellationToken.ThrowIfCancellationRequested () doesn't throw these.

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Tasks/CommentTasksProvider.Legacy.cs [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+			}
+
+			internal static void LoadSolutionContents (Solution sln)
+			{
+				src.Cancel ();
+				src = new CancellationTokenSource ();
+				var token = src.Token;
+
+				Task.Run (delegate {
+					sln.SolutionItemAdded += delegate (object sender, SolutionItemChangeEventArgs e) {
```
Comment (mkrueger) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Tasks/CommentTasksProvider.Legacy.cs]: Is that removed somewhere ? Just adding to a solution event can't it leak ?

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Tasks/CommentTasksProvider.Legacy.cs [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+					return;
+				ProjectCommentTags tags;
+				if (!projectTags.TryGetValue (project, out tags))
+					return;
+				var token = src.Token;
+				var file = doc.FileName;
+				Task.Run (async () => {
+					try {
+						tags.UpdateTags (project, file, await pd.GetTagCommentsAsync (token).ConfigureAwait (false));
+					} catch (TaskCanceledException) {
```
Comment (mkrueger) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Tasks/CommentTasksProvider.Legacy.cs]: OperationCanceledException

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Editor.Extension/TagCommentsTextEditorExtension.cs (hunk 1 above)
Comment (Therzok) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Editor.Extension/TagCommentsTextEditorExtension.cs]: This was copy pasted. I'll switch.

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Tasks/CommentTasksProvider.Legacy.cs (hunk 2 above)
Comment (Therzok) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Tasks/CommentTasksProvider.Legacy.cs]: Copy pasted.

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Tasks/CommentTasksProvider.Legacy.cs (hunk 3 above)
Comment (Therzok) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Tasks/CommentTasksProvider.Legacy.cs]: Will test to see.

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Tasks/CommentTasksProvider.Legacy.cs (hunk 4 above)
Comment (Therzok) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Tasks/CommentTasksProvider.Legacy.cs]: Will change.

**Comment by Therzok** [comment]
@DavidKarlas @mkrueger can you give this a quick test run?

Testing instructions are here: https://devdiv.visualstudio.com/DevDiv/_workitems/edit/568079

**Comment by Therzok** [comment]
TODO: Unit tests

**Comment by Therzok** [comment]
Urghhhh, I need to rebase.

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Tasks/CommentTasksProvider.Legacy.cs (hunk 3 above)
Comment (Therzok) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Tasks/CommentTasksProvider.Legacy.cs]: Won't leak. It doesn't have any reference to an object that would keep it in memory.

In any case, this is a static class, so it'll create a closure which captures the cancellation token. So, it's pretty okay.

**Comment by Therzok** [comment]
Rebased and added a bunch of tests

**Comment by Therzok** [comment]
@DavidKarlas @mkrueger can you please test this and have a look?

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Editor.Extension/TagCommentsTextEditorExtension.cs [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+		protected virtual void OnTasksUpdated (EventArgs e) => TasksUpdated?.Invoke (this, e);
+		public ImmutableArray<QuickTask> QuickTasks => tasks;
+
+		void OnTodoListUpdated (object sender, TodoItemsUpdatedArgs args)
+		{
+			src.Cancel ();
+			src = new CancellationTokenSource ();
+			var token = src.Token;
+			// TODO: If document is null, wait for document changed and query.
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Editor.Extension/TagCommentsTextEditorExtension.cs]: Not sure if this is needed.

**Comment by Therzok** [comment]
The test failure for https://github.com/mono/monodevelop/pull/3996/commits/e2f557160406d12c2f410a578073cab4ad57416b is weird. @slluis any thoughts?

**Comment by Therzok** [comment]
Not sure how to hook a progress monitor to figure out why the test failed. I can't seem to reproduce it locally.

**Comment by Therzok** [comment]
Test fixture successfully run locally for me, seems VSTS has this:
```
2018-04-22T21:26:44.5765140Z System.Runtime.Remoting.RemotingException: Cannot create channel sink to connect to URL a6d11e69_1352_4f46_9b04_44dcebdd3304/1f3c4077_5.rem. An appropriate channel has probably not been registered.
2018-04-22T21:26:44.5839040Z   at System.Runtime.Remoting.RemotingServices.GetClientChannelSinkChain (System.String url, System.Object channelData, System.String& objectUri) [0x00030] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-02/external/bockbuild/builds/mono-x64/mcs/class/corlib/System.Runtime.Remoting/RemotingServices.cs:704 
2018-04-22T21:26:44.5967940Z   at System.Runtime.Remoting.RemotingServices.GetOrCreateClientIdentity (System.Runtime.Remoting.ObjRef objRef, System.Type proxyType, System.Object& clientProxy) [0x00017] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-02/external/bockbuild/builds/mono-x64/mcs/class/corlib/System.Runtime.Remoting/RemotingServices.cs:646 
2018-04-22T21:26:44.6103740Z   at System.Runtime.Remoting.RemotingServices.GetRemoteObject (System.Runtime.Remoting.ObjRef objRef, System.Type proxyType) [0x00000] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-02/external/bockbuild/builds/mono-x64/mcs/class/corlib/System.Runtime.Remoting/RemotingServices.cs:757 
2018-04-22T21:26:44.6238360Z   at System.Runtime.Remoting.RemotingServices.GetProxyForRemoteObject (System.Runtime.Remoting.ObjRef objref, System.Type classToProxy) [0x0001b] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-02/external/bockbuild/builds/mono-x64/mcs/class/corlib/System.Runtime.Remoting/RemotingServices.cs:751 
2018-04-22T21:26:44.6366320Z   at System.Runtime.Remoting.RemotingServices.Unmarshal (System.Runtime.Remoting.ObjRef objectRef, System.Boolean fRefine) [0x0007a] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-02/external/bockbuild/builds/mono-x64/mcs/class/corlib/System.Runtime.Remoting/RemotingServices.cs:286 
2018-04-22T21:26:44.6487550Z   at System.Runtime.Remoting.RemotingServices.Unmarshal (System.Runtime.Remoting.ObjRef objectRef) [0x00000] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-02/external/bockbuild/builds/mono-x64/mcs/class/corlib/System.Runtime.Remoting/RemotingServices.cs:260 
2018-04-22T21:26:44.6607690Z   at System.Runtime.Remoting.ObjRef.GetRealObject (System.Runtime.Serialization.StreamingContext context) [0x0000f] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-02/external/bockbuild/builds/mono-x64/mcs/class/corlib/System.Runtime.Remoting/ObjRef.cs:243 
2018-04-22T21:26:44.6730800Z   at System.Runtime.Serialization.ObjectManager.ResolveObjectReference (System.Runtime.Serialization.ObjectHolder holder) [0x00009] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-02/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/runtime/serialization/objectmanager.cs:303 
2018-04-22T21:26:44.6853370Z   at System.Runtime.Serialization.ObjectManager.DoFixups () [0x00077] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-02/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/runtime/serialization/objectmanager.cs:931 
2018-04-22T21:26:44.6982130Z   at System.Runtime.Serialization.Formatters.Binary.ObjectReader.Deserialize (System.Runtime.Remoting.Messaging.HeaderHandler handler, System.Runtime.Serialization.Formatters.Binary.__BinaryParser serParser, System.Boolean fCheck, System.Boolean isCrossAppDomain, System.Runtime.Remoting.Messaging.IMethodCallMessage methodCallMessage) [0x00077] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-02/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/runtime/serialization/formatters/binary/binaryobjectreader.cs:181 
2018-04-22T21:26:44.7121290Z   at System.Runtime.Serialization.Formatters.Binary.BinaryFormatter.Deserialize (System.IO.Stream serializationStream, System.Runtime.Remoting.Messaging.HeaderHandler handler, System.Boolean fCheck, System.Boolean isCrossAppDomain, System.Runtime.Remoting.Messaging.IMethodCallMessage methodCallMessage) [0x00099] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-02/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/runtime/serialization/formatters/binary/binaryformatter.cs:197 
2018-04-22T21:26:44.7285560Z   at System.Runtime.Serialization.Formatters.Binary.BinaryFormatter.Deserialize (System.IO.Stream serializationStream, System.Runtime.Remoting.Messaging.HeaderHandler handler, System.Boolean fCheck, System.Runtime.Remoting.Messaging.IMethodCallMessage methodCallMessage) [0x00000] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-02/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/runtime/serialization/formatters/binary/binaryformatter.cs:173 
2018-04-22T21:26:44.7445750Z   at System.Runtime.Serialization.Formatters.Binary.BinaryFormatter.Deserialize (System.IO.Stream serializationStream, System.Runtime.Remoting.Messaging.HeaderHandler handler, System.Boolean fCheck) [0x00000] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-02/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/runtime/serialization/formatters/binary/binaryformatter.cs:118 
2018-04-22T21:26:44.7608840Z   at System.Runtime.Serialization.Formatters.Binary.BinaryFormatter.Deserialize (System.IO.Stream serializationStream, System.Runtime.Remoting.Messaging.HeaderHandler handler) [0x00000] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-02/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/runtime/serialization/formatters/binary/binaryformatter.cs:149 
2018-04-22T21:26:44.7792940Z   at System.Runtime.Serialization.Formatters.Binary.BinaryFormatter.Deserialize (System.IO.Stream serializationStream) [0x00000] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-02/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/runtime/serialization/formatters/binary/binaryformatter.cs:111 
2018-04-22T21:26:44.7905330Z   at MonoDevelopProcessHost.Main (System.String[] args) [0x00099] in /Users/vsts/agent/2.131.0/work/1/s/monodevelop/main/src/tools/mdhost/src/mdhost.cs:89 
```

**Comment by Therzok** [comment]
So, tests do not fail locally. I have no clue why they fail on VSTS.

**Comment by Therzok** [comment]
Okay, running again. Seems like I had the wrong env var for testing.

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Editor.Extension/TagCommentsTextEditorExtension.cs [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+			var project = ws.GetMonoProject (args.ProjectId);
+			if (project == null)
+				return;
+
+			var newTasks = ImmutableArray<QuickTask>.Empty.ToBuilder ();
+			foreach (var todoItem in args.TodoItems) {
+				if (token.IsCancellationRequested)
+					return;
+				
+				var offset = Editor.LocationToOffset (todoItem.MappedLine, todoItem.MappedColumn);
```
Comment (mkrueger) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Editor.Extension/TagCommentsTextEditorExtension.cs]: Should run on the UI thread - this is not thread safe.

**Comment by Therzok** [comment]
Fixed UI thread issue.

**Comment by Therzok** [comment]
yay, finally a green build

**Comment by slluis** [comment]
@marius Are we losing any functionality with this change?

**Comment by Therzok** [comment]
C# is working exactly the same now, except todo comments are parsed incrementally in background

**Pull request merged**